### PR TITLE
fix: make `pyarrow` an optional dependency post-3.20.0 yanked release

### LINKED ISSUE #1877: pyarrow import error in 3.20.0
#### Environment details

  - OS type and version: MacOS 14.4.1 (23E224)
  - Python version: 3.10
  - pip version: 24.0
  - `google-cloud-bigquery` version: 3.20.0 

#### Steps to reproduce
Create a new virtualenv, install google-cloud-bigquery==3.20.0, attempt to `import google.cloud.bigquery`

```
$ virtualenv -p python3.10 venv
created virtual environment CPython3.10.14.final.0-64 in 322ms
  creator CPython3macOsBrew(dest=/Users/me/venv, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/Users/me/Library/Application Support/virtualenv)
    added seed packages: pip==24.0, setuptools==69.1.1, wheel==0.42.0
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator
```  
 
```
$ source venv/bin/activate
```
```
$ pip install -i https://pypi.python.org/simple google-cloud-bigquery==3.20.0
Looking in indexes: https://pypi.python.org/simple
Collecting google-cloud-bigquery==3.20.0
  Using cached google_cloud_bigquery-3.20.0-py2.py3-none-any.whl.metadata (8.9 kB)
Collecting google-api-core!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1 (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-bigquery==3.20.0)
  Using cached google_api_core-2.18.0-py3-none-any.whl.metadata (2.7 kB)
Collecting google-auth<3.0.0dev,>=2.14.1 (from google-cloud-bigquery==3.20.0)
  Using cached google_auth-2.29.0-py2.py3-none-any.whl.metadata (4.7 kB)
Collecting google-cloud-core<3.0.0dev,>=1.6.0 (from google-cloud-bigquery==3.20.0)
  Using cached google_cloud_core-2.4.1-py2.py3-none-any.whl.metadata (2.7 kB)
Collecting google-resumable-media<3.0dev,>=0.6.0 (from google-cloud-bigquery==3.20.0)
  Using cached google_resumable_media-2.7.0-py2.py3-none-any.whl.metadata (2.2 kB)
Collecting packaging>=20.0.0 (from google-cloud-bigquery==3.20.0)
  Using cached packaging-24.0-py3-none-any.whl.metadata (3.2 kB)
Collecting python-dateutil<3.0dev,>=2.7.2 (from google-cloud-bigquery==3.20.0)
  Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting requests<3.0.0dev,>=2.21.0 (from google-cloud-bigquery==3.20.0)
  Using cached requests-2.31.0-py3-none-any.whl.metadata (4.6 kB)
Collecting googleapis-common-protos<2.0.dev0,>=1.56.2 (from google-api-core!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-bigquery==3.20.0)
  Using cached googleapis_common_protos-1.63.0-py2.py3-none-any.whl.metadata (1.5 kB)
Collecting protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0.dev0,>=3.19.5 (from google-api-core!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-bigquery==3.20.0)
  Using cached protobuf-4.25.3-cp37-abi3-macosx_10_9_universal2.whl.metadata (541 bytes)
Collecting proto-plus<2.0.0dev,>=1.22.3 (from google-api-core!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-bigquery==3.20.0)
  Using cached proto_plus-1.23.0-py3-none-any.whl.metadata (2.2 kB)
Collecting grpcio<2.0dev,>=1.33.2 (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-bigquery==3.20.0)
  Using cached grpcio-1.62.1-cp310-cp310-macosx_12_0_universal2.whl.metadata (4.0 kB)
Collecting grpcio-status<2.0.dev0,>=1.33.2 (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-bigquery==3.20.0)
  Using cached grpcio_status-1.62.1-py3-none-any.whl.metadata (1.3 kB)
Collecting cachetools<6.0,>=2.0.0 (from google-auth<3.0.0dev,>=2.14.1->google-cloud-bigquery==3.20.0)
  Using cached cachetools-5.3.3-py3-none-any.whl.metadata (5.3 kB)
Collecting pyasn1-modules>=0.2.1 (from google-auth<3.0.0dev,>=2.14.1->google-cloud-bigquery==3.20.0)
  Using cached pyasn1_modules-0.4.0-py3-none-any.whl.metadata (3.4 kB)
Collecting rsa<5,>=3.1.4 (from google-auth<3.0.0dev,>=2.14.1->google-cloud-bigquery==3.20.0)
  Using cached rsa-4.9-py3-none-any.whl.metadata (4.2 kB)
Collecting google-crc32c<2.0dev,>=1.0 (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery==3.20.0)
  Using cached google_crc32c-1.5.0-cp310-cp310-macosx_10_9_universal2.whl.metadata (2.3 kB)
Collecting six>=1.5 (from python-dateutil<3.0dev,>=2.7.2->google-cloud-bigquery==3.20.0)
  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Collecting charset-normalizer<4,>=2 (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery==3.20.0)
  Using cached charset_normalizer-3.3.2-cp310-cp310-macosx_11_0_arm64.whl.metadata (33 kB)
Collecting idna<4,>=2.5 (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery==3.20.0)
  Using cached idna-3.6-py3-none-any.whl.metadata (9.9 kB)
Collecting urllib3<3,>=1.21.1 (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery==3.20.0)
  Using cached urllib3-2.2.1-py3-none-any.whl.metadata (6.4 kB)
Collecting certifi>=2017.4.17 (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery==3.20.0)
  Using cached certifi-2024.2.2-py3-none-any.whl.metadata (2.2 kB)
Collecting pyasn1<0.7.0,>=0.4.6 (from pyasn1-modules>=0.2.1->google-auth<3.0.0dev,>=2.14.1->google-cloud-bigquery==3.20.0)
  Using cached pyasn1-0.6.0-py2.py3-none-any.whl.metadata (8.3 kB)
Using cached google_cloud_bigquery-3.20.0-py2.py3-none-any.whl (233 kB)
Using cached google_api_core-2.18.0-py3-none-any.whl (138 kB)
Using cached google_auth-2.29.0-py2.py3-none-any.whl (189 kB)
Using cached google_cloud_core-2.4.1-py2.py3-none-any.whl (29 kB)
Using cached google_resumable_media-2.7.0-py2.py3-none-any.whl (80 kB)
Using cached packaging-24.0-py3-none-any.whl (53 kB)
Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
Using cached requests-2.31.0-py3-none-any.whl (62 kB)
Using cached cachetools-5.3.3-py3-none-any.whl (9.3 kB)
Using cached certifi-2024.2.2-py3-none-any.whl (163 kB)
Using cached charset_normalizer-3.3.2-cp310-cp310-macosx_11_0_arm64.whl (120 kB)
Using cached google_crc32c-1.5.0-cp310-cp310-macosx_10_9_universal2.whl (32 kB)
Using cached googleapis_common_protos-1.63.0-py2.py3-none-any.whl (229 kB)
Using cached grpcio-1.62.1-cp310-cp310-macosx_12_0_universal2.whl (10.0 MB)
Using cached grpcio_status-1.62.1-py3-none-any.whl (14 kB)
Using cached idna-3.6-py3-none-any.whl (61 kB)
Using cached proto_plus-1.23.0-py3-none-any.whl (48 kB)
Using cached protobuf-4.25.3-cp37-abi3-macosx_10_9_universal2.whl (394 kB)
Using cached pyasn1_modules-0.4.0-py3-none-any.whl (181 kB)
Using cached rsa-4.9-py3-none-any.whl (34 kB)
Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Using cached urllib3-2.2.1-py3-none-any.whl (121 kB)
Using cached pyasn1-0.6.0-py2.py3-none-any.whl (85 kB)
Installing collected packages: urllib3, six, pyasn1, protobuf, packaging, idna, grpcio, google-crc32c, charset-normalizer, certifi, cachetools, rsa, requests, python-dateutil, pyasn1-modules, proto-plus, googleapis-common-protos, google-resumable-media, grpcio-status, google-auth, google-api-core, google-cloud-core, google-cloud-bigquery
Successfully installed cachetools-5.3.3 certifi-2024.2.2 charset-normalizer-3.3.2 google-api-core-2.18.0 google-auth-2.29.0 google-cloud-bigquery-3.20.0 google-cloud-core-2.4.1 google-crc32c-1.5.0 google-resumable-media-2.7.0 googleapis-common-protos-1.63.0 grpcio-1.62.1 grpcio-status-1.62.1 idna-3.6 packaging-24.0 proto-plus-1.23.0 protobuf-4.25.3 pyasn1-0.6.0 pyasn1-modules-0.4.0 python-dateutil-2.9.0.post0 requests-2.31.0 rsa-4.9 six-1.16.0 urllib3-2.2.1
```

```
$ python -c "import google.cloud.bigquery"
Traceback (most recent call last):
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/_versions_helpers.py", line 75, in try_import
    import pyarrow
ModuleNotFoundError: No module named 'pyarrow'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/__init__.py", line 35, in <module>
    from google.cloud.bigquery.client import Client
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 69, in <module>
    from google.cloud.bigquery import _job_helpers
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/_job_helpers.py", line 47, in <module>
    from google.cloud.bigquery import job
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/job/__init__.py", line 27, in <module>
    from google.cloud.bigquery.job.copy_ import CopyJob
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/job/copy_.py", line 22, in <module>
    from google.cloud.bigquery.table import TableReference
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/table.py", line 62, in <module>
    from google.cloud.bigquery import _pandas_helpers
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/_pandas_helpers.py", line 52, in <module>
    pyarrow = _versions_helpers.PYARROW_VERSIONS.try_import(raise_if_error=True)
  File "/Users/me/venv/lib/python3.10/site-packages/google/cloud/bigquery/_versions_helpers.py", line 78, in try_import
    raise exceptions.LegacyPyarrowError(
google.cloud.bigquery.exceptions.LegacyPyarrowError: pyarrow package not found. Install pyarrow version >= 3.0.0.
```

### PR DESCRIPTION
**Pull request by tswast**
Partial rollback of https://github.com/googleapis/python-bigquery/commit/0ac6e9bf186945832f5dcdf5a4d95667b4da223e

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #1877 🦕

### CODE REVIEW

File: google/cloud/bigquery/_pandas_helpers.py [hunk 1]
```diff
@@ -309,10 +308,10 @@ def bq_to_arrow_array(series, bq_field):
         if field_type_upper in schema._STRUCT_TYPES:
             return pyarrow.StructArray.from_pandas(series, type=arrow_type)
         return pyarrow.Array.from_pandas(series, type=arrow_type)
-    except ArrowTypeError:  # pragma: NO COVER
+    except pyarrow.ArrowTypeError:  # pragma: NO COVER
```
Comment (leahecole) [inline google/cloud/bigquery/_pandas_helpers.py:311]: Do we want to keep the `# pragma: NO COVER` Here?

Comment (tswast) [reply google/cloud/bigquery/_pandas_helpers.py:311]: Ideally not. We really should have a test covering this case. I'll see what I can do.

Comment (tswast) [reply google/cloud/bigquery/_pandas_helpers.py]: Just did a big `# pragma: NO COVER` sweep and was able to remove a ton! One straggler with `shapely` where I've commented the reason why.

**Comment by chalmerlowe** [comment]
@tswast

We appear to have two conflicting versions of `pyarrow`:

The `noxfile.py` file identifies these versions of Python:

Line 40: `UNIT_TEST_PYTHON_VERSIONS = ["3.7", "3.8", "3.12"]`

Later when we reference 

Line 125: `UNIT_TEST_PYTHON_VERSIONS[0]`

we are implicitly referencing version `3.7` (at least until the definition of this constant changes) AND we then tell the system to install version 1.0.0 of pyarrow.
Line 126: `session.install("pyarrow==1.0.0")`

But in `testing/constraints-3.7.txt` we list the pyarrow version as 

Line 30: `pyarrow==3.0.0`

QUESTION: should those two match?
QUESTION: if we are making `pyarrow` optional, why in the `unit_noextras` session are we requiring `pyarrow`?

**Comment by tswast** [comment]
> QUESTION: should those two match?

No. They should not. `testing/constraints-3.7.txt` covers when pyarrow is installed as part of our package installation, e.g. `unit-3.7`. This doesn't apply to the `unit_noextras-3.7` session.

> QUESTION: if we are making pyarrow optional, why in the unit_noextras session are we requiring pyarrow?

There is a comment explaining exactly this. See: https://github.com/googleapis/python-bigquery/issues/933 Requiring the correct version of pyarrow to be installed broke some environments where pyarrow was stuck on an older version. We need to be robust to this failure mode.

**Comment by tswast** [comment]
See: https://pip.pypa.io/en/stable/user_guide/#constraints-files for more information on constraints files. When applied, it affects what version of a package is installed but it doesn't affect packages that aren't included because of a transitive dependency of the package being installed.

**Review by Linchin — APPROVED** [review]
LGTM. I will not merge for now, waiting for Chalmer to see the response.

**Comment by chalmerlowe** [comment]
Thanks @tswast 
Took a look at the write-up on `constraints.txt` files.
My understanding is:
* unlike a `requirements.txt` file, it does NOT require that a specific package be installed
* HOWEVER, if that package DOES get installed (maybe as a dependency of something else OR as an optional package), it is constrained to what is allowed in the `constraints.txt` file.

**Comment by tswast** [comment]
That's correct. Note that we are installing `pyarrow` without a constraints file in `unit_noextras` and `pyarrow` is not included as a dependency in any subsequent installations, so the constraints do not apply.

**Review by chalmerlowe — APPROVED** [review]
LGTM

**Review by shollyman — APPROVED** [review]
Generally, looks like a good unwinding of the behavior changes and test configurations.  I did add one question around test philosophy that may be outside the scope of this PR specifically.

File: tests/unit/test__versions_helpers.py [hunk 2]
```diff
@@ -62,6 +59,16 @@ def test_try_import_raises_error_w_legacy_pyarrow():
             versions.try_import(raise_if_error=True)
 
 
+@pytest.mark.skipif(
+    pyarrow is not None,
+    reason="pyarrow is installed, but this test needs it not to be",
+)
+def test_try_import_raises_error_w_no_pyarrow():
```
[the remark this answers is not in the document]
Comment (tswast) [reply tests/unit/test__versions_helpers.py:66]: Coverage should catch that. We check coverage on test files as well as the core library.

[the remark this answers is not in the document]
Comment (tswast) [reply tests/unit/test__versions_helpers.py:66]: See: https://github.com/googleapis/python-bigquery/blob/7dfee0c585d2a3781ffc6e769c7c8bbe4dbe9714/noxfile.py#L97

**Pull request merged**